File: permutation.py
#!/usr/bin/env python
#--*--coding:utf8--*--
'''要修改，使得空间是O(1)的，'''
global permu
permu=[]
def _permutation(invertation):
    '''Construct a permutation from 1 to N,by the invertation table of
        a permutation.'''
    if(isinstance(invertation, list)):
        length=len(invertation)
        permu=[]
        for i in range(0, length):
            permu.insert(invertation[i], length-i)
#        print permu
    else:
        print "Type error"

def gerninvertation(scale):
    '''Construct a inversed table generator.The table's scale is determinded by
    argument 'scale'.'''
    invertation=[]
    for i in range(scale):
        invertation.append(0)
    yield invertation
    try:
        while True:
            invertation=add(invertation,1)
            yield invertation
    except FlowOverError:
        print "It's over.!"


#    limit=reduce(lambda x,y:x*y,range(1,scale+1))
#    print limit
#    i=1
#    while i<=limit:
#        if i==1:
#            i=i+1
#
#        else:
#            try:
#                invertation=add(invertation,1)
#            except FlowOverError:
#                print "It's over.!"
#            i=i+1
#            yield invertation


def add(invertation,lens=1):
    '''An inversed table is seen as a number,but on the i-th bit i-th number
    system,we emulate the add operation.'''
    length=len(invertation)

#we need the following gaurd variable.The 'flag' marks the state whether we
#should stop calculating.'index' mark which bit we are.forward marks how much
#we should carrying forward.
    flag=0
    index=0
    forward=1
#if addition has done (flag==0),stop.
    while flag==0 and (index < length):

        if invertation[index]+forward < index+1:
            invertation[index]=invertation[index]+forward
            flag=1
        else:
            a, b=divmod(invertation[index]+forward,index+1)
            invertation[index]=b
            forward=a
            index = index+1
    if flag==0:
        raise FlowOverError
    else:
        return invertation

def permutation(scale):
    generator=gerninvertation(scale)
    try:
        while True:
            invertation=generator.next()
            _permutation(invertation)
    except StopIteration:
        print "permutation has print over."
        return 0

class FlowOverError(BaseException):
    def __init__(self):
        print "FlowOverError"

if __name__=='__main__':
    permutation(23)

File: permutation.c
#include <stdio.h>
#include "utility.h"
/*This is a permutation generator.
 */
typedef struct permu{
    int *permutation;
    int *inversetable;
    int length;
    int overflow;
} permu;


int _add(int *invertation,int length)
{
    /* This function is modify the data in the place.
     * 'flag' is a mark of whether the operation has done.
     * 'index' is the place where the carrying of number take place.
     * 'forward' is the amount of carrying.
     */
    int flag = 0,index = 0,forward = 1;
    while(flag==0 && index < length){
        /*no carrying*/
        if(*(invertation+index)+forward < index +1){
            *(invertation +index)=*(invertation+index)+forward;
            flag = 1;
        }
        /*carry the number*/
        else{
            forward = (int)((*(invertation+index) + forward)/(index + 1));
            *(invertation+index) = (int)((*(invertation+index) + forward)%(index + 1));
            index++;
        }
    }
    if(flag == 0) {
        int i;
        for(i=0;i<length;i++){
            *(invertation+i)=-1;
        }
        return 0;
    }
    else return 1;
}


int _geninversetable(permu *container)
{
    int *inversetable = container->inversetable;
    int length = container->length;

    if(_add(inversetable,length)){
        return 1;
    }
    else{
        container->overflow = 1;
        return 0;
    }
}

/*Produce a permutation from a inversetable.Store the permutation in permutation variable.
 * all this is in the structure permu.*/
int _permutation(permu *permutation)
{
    int *permu = permutation->permutation;
    int *inversetable = permutation->inversetable;
    int length = permutation->length;

    int i;
    int insert,inverseNum;
    for(i=0;i< length;i++){
        insert = length - i;/*The number will be inserted into the array.*/
        inverseNum = *(inversetable + i);/*The amount how many numbers bigger before.*/

        _rshift(permu,length,inverseNum,i);/*Empty the place where we want to insert number.*/
        *(permu+inverseNum) = insert;/*Insert.*/
    }

    return 1;
}

int permutation(permu *p){
    while(!(p->overflow)){
        _geninversetable(p);
        _permutation(p);
        print_1array(p->permutation,p->length);
    }
    return 1;
}
int main(int argc, const char *argv[])
{
#ifdef TEST
    permu p;
    int permutations[26];
    int inversetable[26];
    _initialize(inversetable,26,0);    

    /*Initialize the structure.*/
    p.permutation = permutations;
    p.inversetable = inversetable;
    p.length = 26;
    p.overflow = 0;
    
    permutation(&p);
#endif 
    return 0;
}
